src/tree/updater_quantile_hist.cc

```diff
@@ -554,40 +562,77 @@ void QuantileHistMaker::Builder::InitData(const GHistIndexMatrix& gmat,
   builder_monitor_.Stop("InitData");
 }
 
-void QuantileHistMaker::Builder::EvaluateSplit(const int nid,
+// if sum of statistics for non-missing values in the node
```

Comment:
@hcho3 I thought missing value is removed during construction of DMatrix?

Reply:
@trivialfis DMatrix does not store missing values physically, but the split-finding algorithm still needs to take account of missing values.

Explanation for this piece of function: Let `NM` represent the set of all data points whose (particular) feature `f` is not missing, and let `S` represent the set of data points associated with the particular node `snode`. Then we want to ask whether two sets `intersection(NM, S)` and `S` are identical or not. There are two possibilities:

1. `intersection(NM, S) == S`. That is, `S` is a subset of `NM` and all data points in `S` has a non-missing value for the particular feature `f`.
2. `intersection(NM, S) != S`. In this case, there exists one data point in `S` which has a missing value for feature `f`.

Reply:
@hcho3 Thanks for the detailed explanation.  I forgot this optimization.